plutus-contract-exe/src/Language/Plutus/Contract.hs

```diff
@@ -0,0 +1,106 @@
+{-
+    Event-based interface between contract executables and the app platform and
+    (by extension) the wallet.
+
+    Two types of events are defined:
+
+    1. 'ContractOut'. Events produced by the contract for consumption by app
+       platform and wallet. Includes transactions and instructions to start
+       watching interesting addresses on the ledger.
+
+    2. 'LedgerUpdate'. Events that inform the contract about changes to the
+       ledger state.
+
+    Contracts will offer an HTTP interface with the following routes.
+
+    * 1 route per contract endpoint (as defined by contract author)
+    * 1 route for submitting 'LedgerUpdate' events
+
+    NOTE: With this design, there are two classes of input events, only one of
+          which has a proper data type: Input events from the ledger (in the
+          form of 'LedgerUpdate') and input events from the user (in the form
+          of HTTP endpoints).
+
+          In my opinion it would be cleaner to add a
+
+          'ContractEndpoint EndpointArgs'
+
+          constructor to 'LedgerUpdate' (and rename the type) that captures all
+          contract endpoints activated by the user. I think this will make
+          testing a lot easier. But we need to think about the 'EndpointArgs'
+          type and how it maps to the actual contract endpoints.
+-}
+module Language.Plutus.Contract(
+    ContractOut(..),
+    LedgerUpdate(..)
+    ) where
+
+import qualified Data.Aeson   as Aeson
+import           Data.Text    (Text)
+import           GHC.Generics (Generic)
+
+import qualified Ledger.Types as T
+
+-- | Events that are produced by contract endpoints.
+data ContractOut =
+      SubmitTransaction T.Tx
+      -- ^ Produce a transaction. The transaction is unsigned and may or may
+      --   not be balanced. If it is not balanced then the wallet is expected to
+      --   balance it by
+      --   (a) adding public key inputs if the transaction produces more than
+      --       it spends
+      --   (b) adding public key outputs if the transaction spends more than it
+      --       produces
+
+      | StartWatching T.Address
+      -- ^ Start watching an address. This adds the address to the set of
+      --   "interesting addresses" of this contract instance.
+
+      -- | Checkpoint Aeson.Value
+      -- ^ Write a checkpoint (a JSON value)
+      --   Commented out for now. We need to think about how best to reset the
+      --   contract state to a checkpoint. Maybe there could be a POST /state
+      --   endpoint?
+      --   Another option would be to allow POST and GET at the /state route
+      --   (instead of the 'Checkpoint' constructor for 'ContractOut'). This
+      --   would allow the app platform greater scope in starting and stopping
+      --   the contract process.
+
+      | ContractError Text
+      -- ^ An error occurred during contract execution.
+      --   NOTE: Should we also set the appropriate HTTP status code?
+
+      | ContractFinished
```

Comment:
I'm also slightly confused by this.

Reply:
It's to allow the contract backend to delete all triggers it has registered for this contract